### ISSUE TITLE
Docker processes starting up with unique api keys

### ISSUE THREAD

**Issue by dylanmcreynolds** [bluesky/tiled#320]
Adding this as it's own issue, referencing #317.

Starting up the server following [this](https://github.com/bluesky/tiled/blob/main/docs/source/how-to/docker.md).

- First, I created the config.yml file and created an example_files directory
- Then I pulled the image
- Finally, I ran the command:
```
docker run --rm -p 8000:8000 \
  --mount type=bind,source="$(pwd)/example_files",target=/deploy/example_files \
  --mount type=bind,source="$(pwd)/config.yml",target=/deploy/config.yml \
  --env TILED_CONFIG=/deploy/config.yml  ghcr.io/bluesky/tiled:main
```

When the tiled processes started up, a couple strange things pop up in the log below. First, the message says it's in "public mode" yet still runs with api_keys. 

Second, the api_key in each process is different, meaning that the client really has no way to consistently  talk to the processes through gunicorn.

```
Using configuration from /deploy/config.yml
Using configuration from /deploy/config.yml
Using configuration from /deploy/config.yml
Using configuration from /deploy/config.yml

    Tiled server is running in "public" mode, permitting open, anonymous access
    for reading. Any data that is not specifically controlled with an access
    policy will be visible to anyone who can connect to this server.


    Navigate a web browser to:

    http://0.0.0.0:8000?api_key=488691f38b12bfc7939c3faaa4acd8fec3547423f2d867d8fc4983d06fc9c118

    or connect a Tiled client to:

    http://0.0.0.0:8000/api?api_key=488691f38b12bfc7939c3faaa4acd8fec3547423f2d867d8fc4983d06fc9c118



    Tiled server is running in "public" mode, permitting open, anonymous access
    for reading. Any data that is not specifically controlled with an access
    policy will be visible to anyone who can connect to this server.


    Navigate a web browser to:

    http://0.0.0.0:8000?api_key=f1740e7cbd22799d4e2a4bfe94dabbd34d98e93017554303e5a28b9a5c0c4f5a

    or connect a Tiled client to:

    http://0.0.0.0:8000/api?api_key=f1740e7cbd22799d4e2a4bfe94dabbd34d98e93017554303e5a28b9a5c0c4f5a



    Tiled server is running in "public" mode, permitting open, anonymous access
    for reading. Any data that is not specifically controlled with an access
    policy will be visible to anyone who can connect to this server.


    Navigate a web browser to:

    http://0.0.0.0:8000?api_key=9ef650abb151c900010553210af5cda989bbe7b4bae16e3b8d433d294d158c8f

    or connect a Tiled client to:

    http://0.0.0.0:8000/api?api_key=9ef650abb151c900010553210af5cda989bbe7b4bae16e3b8d433d294d158c8f



    Tiled server is running in "public" mode, permitting open, anonymous access
    for reading. Any data that is not specifically controlled with an access
    policy will be visible to anyone who can connect to this server.


    Navigate a web browser to:

    http://0.0.0.0:8000?api_key=d745f980e17a75de581003cff4a2b5db2b13160fb2f15396d0bbdbcc2af34f68

    or connect a Tiled client to:

    http://0.0.0.0:8000/api?api_key=d745f980e17a75de581003cff4a2b5db2b13160fb2f15396d0bbdbcc2af34f68

```

**Comment by kleinhenz** [comment]
I think the only problem is that it prints an api key even in public mode. The actual behavior is as expected. You can connect without an api key and if you change the config to `allow_anonymous_access: false` it will correctly complain that you have not specified an api key.

**Comment by dylanmcreynolds** [comment]
I'm not sure this is true. When we were debugging this with Tanny, she configured the client to use one api key, and about a quarter of the messages succeeded, and a quarter failed.

**Comment by kleinhenz** [comment]
it seems to work correctly for me:
```
trees:
  - path: /
    tree: tiled.adapters.files:DirectoryAdapter.from_directory
    args:
      directory: "example_files"
uvicorn:
  host: 0.0.0.0
  port: 8000
authentication:
  allow_anonymous_access: false
  single_user_api_key: abc
```

```
Using configuration from /deploy/config.yml
Using configuration from /deploy/config.yml
Using configuration from /deploy/config.yml
Using configuration from /deploy/config.yml

    Navigate a web browser to:

    http://0.0.0.0:8000?api_key=abc

    or connect a Tiled client to:

    http://0.0.0.0:8000/api?api_key=abc



    Navigate a web browser to:

    http://0.0.0.0:8000?api_key=abc

    or connect a Tiled client to:

    http://0.0.0.0:8000/api?api_key=abc



    Navigate a web browser to:

    http://0.0.0.0:8000?api_key=abc

    or connect a Tiled client to:

    http://0.0.0.0:8000/api?api_key=abc



    Navigate a web browser to:

    http://0.0.0.0:8000?api_key=abc

    or connect a Tiled client to:

    http://0.0.0.0:8000/api?api_key=abc
```

**Comment by dylanmcreynolds** [comment]
And without specifying  `single_user_api_key`?

**Comment by kleinhenz** [comment]
```
When Tiled is configured for single-user access (i.e. without an
Authenticator), it must either be set to allow anonymous (public) access like

authentication:
  allow_anonymous_access: true
  ...

or else a single-user API key must be provided via configuration like

authentication:
  single_user_api_key: SECRET
  ...

or via the environment variable TILED_SINGLE_USER_API_KEY. Exiting...
```

which  think is the correct behavior.

**Comment by dylanmcreynolds** [comment]
Yes. What I found is that if you startup in Docker with just `allow_anonymous_access`, and no `single_user_api_key`, (following the example [in](https://github.com/bluesky/tiled/blob/main/docs/source/how-to/docker.md)   the workers start up with individual API keys.

**Comment by kleinhenz** [comment]
closed by #322. The resolution is to require `single_user_api_key` to be explicitly set even if `allow_anonymous_access` is true. This makes sense because even with `allow_anonymous_access` there may still be actions (e.g. writing, admin) that require authentication.
